xarray/core/common.py

```diff
@@ -610,21 +610,25 @@ def where(self, cond, other=None, drop=False):
         if drop:
             from .dataarray import DataArray
             from .dataset import Dataset
+            from .alignment import align
+
+            if not isinstance(cond, (Dataset, DataArray)):
+                raise TypeError("Cond argument is %r but must be a %r or %r" %
+                                (cond, Dataset, DataArray))
+            # align so we can use integer indexing
+            self, cond = align(self, cond)
+
             # get cond with the minimal size needed for the Dataset
             if isinstance(cond, Dataset):
                 clipcond = cond.to_array().any('variable')
-            elif isinstance(cond, DataArray):
-                clipcond = cond
             else:
-                raise TypeError("Cond argument is %r but must be a %r or %r" %
-                                (cond, Dataset, DataArray))
+                clipcond = cond
 
             # clip the data corresponding to coordinate dims that are not used
-            clip = dict(zip(clipcond.dims, [np.unique(adim)
-                                            for adim in np.nonzero(clipcond.values)]))
-            outcond = cond.isel(**clip)
-            indexers = {dim: outcond.get_index(dim) for dim in outcond.dims}
```

Comment:
It worked, but it filled in a default index based on the length of the dimension -- which isn't doing the right thing if we clipped values.